syntax = "proto3";

package masterworker;

service MapperReducer {
	rpc MapCall (MapRequest) returns (MapReply) {}
	rpc ReduceCall (ReduceRequest) returns (ReduceReply) {}
}

message MapRequest {
	uint32 num_reducers = 1;
	repeated ShardInfo shards = 2;
}

message ShardInfo {
	uint64 begin = 1;
	uint64 end = 2;
	string file_name = 3;
}

message MapReply {
	repeated MapResults results = 1;
}

message MapResults {
	string key_tag = 1;
	string file_name = 2;
}

message ReduceRequest {
	string key_tag = 1;
	string out_file = 2;
	repeated ReduceBatch batch = 3;
}

message ReduceBatch {
	string in_file = 1;
}

message ReduceReply {
	bool isComplete = 1;
}
